### ISSUE TITLE
Clarify meaning of Security Domain

### ISSUE THREAD

**Issue by hannestschofenig** [ietf-teep/architecture#7]
The term "security domain" is described in the architecture document as: 

> A Security Domain (SD) concept is used as the security boundary
inside a TEE for trusted applications.

There seems to be an implicit assumptions that trusted applications do not share resources with other trusted applications. The security domain concept then seems to allow to enable this sharing. Global Platform has been referenced as the source of this concept but unfortunately I wasn't able to find a definition in the Global Platform documents. The question is also what resources would be shared between TAs. 

The introduction of the Security Domain concept has also other side effects on the protocol specification: 

First, it requires separate messaging for creating and deleting security domains. 

Second, it requires added protocol elements to reference security domains. 

Third, the Anonymous Attestation Key (AAK) is established and maintained for use with security domains.

**Comment by dthaler** [comment]
I think some TEEs may not have a concept of shared resources (i.e., only support 1 TA per SD).  So there should be an efficient way to use OTrP for such devices that does not incur extra overhead.

**Comment by hannestschofenig** [comment]
This topic got discussed at the virtual interim meeting (22nd Oct.2018): 

> DaveT: if TEE just tracks which TAM installed a TA, it can only accept changes for that TA from that TAM (without needing a SD concept, it can just treat that TAM as the owner).  If a TAM wants to keep track of SD concept inside the TAM, without telling the TEE, it could do so internally.

**Comment by dmwheel1** [comment]
Question: Is there an additional security and/or privacy issue here?
If TAM key is used to manage all installations, then a single exposure or break of the TAM key results in the loss of security of all the TAMs installed applications (break one key, get access to everything). By putting applications into SD's, we distribute the cryptographic dependence onto other keys. Of course a compromise of a TAM could still expose all SD keys, but this depends on how the TAM is set up.

Additionally, depending on how the book-keeping is done, SDs can provide some privacy as to which TAM manages them, if just the key fingerprint is used, and different keys are used for different SDs. This all depends on how the TAM interfaces with the protocol and whether it generates new keys for each new SD.

I am still a bit concerned with the overhead messaging to create/manage SDs.
But perhaps we should consider allowing management of an application be assigned to an arbitrary key.

**Comment by hannestschofenig** [comment]
DavidW suggested to look at the management concept where multiple binaries are pushed to the TEE via the help of a manifest and then those are installed in batch without having to issue separate message exchanges for each messages. You would create a key for these TAs and would then manage these apps with the helps of these keys.

**Comment by hannestschofenig** [comment]
At IETF#103 David Thaler described four different cases, namely 
1) Security Domain concept is mandatory to implement and to use 
2) Additional roundtrip for creating the Security Domain is optional. AAK is still mandatory to use. 
3) Security Domain as well as the  Anonymous Attestation Key (AAK) concepts are optional
4) Security Domain concept is removed from the spec

**Comment by hannestschofenig** [comment]
Additional notes from the IETF #103 meeting minutes: 

Issue #7 - clarify meaning of Security Domain; may need to be supplemented with
description of how to address the management use-case - specifically, could it
sensibly be addressed by defining
1 - mandatory security domain
2 - optional security domain, but with a requirement to provision a public key per
Security Domain] -- this is what the doc attempts to define currently.
3 - security domain optional, with possibility of avoiding the public key provisioning
round-trip -- this option is not currently defined/documented
4 - remove concept of SD from the document
Intuition of how the Security Domain works is really needed. Management is a potential
useful case for SDs.
Dave W: agree it's a substantive issue, but it really needs more discussion before
decision.
Dave T/Erik N: this boils down to defining what a Security Domain is really used for; is
it a management component, or an isolation mechanism?

**Comment by hannestschofenig** [comment]
I did some research based on available smart card/GP documents and it seems that even there the concept of a security domain is not well defined. I would therefore suggest to ignore the term and approach the topic differently, namely by looking at what are the desirable features.

I believe it is important that one trusted app cannot just read and modify the state of another trusted app. If two trusted apps interact with each other than they have to use a well-defined API, which would be outside the scope of the protocol mechanisms we are looking into. 

Reviewing the OTrP specification it turns out that the security domain has one distinct purpose, namely to serve as the identifier for the Anonymous Key (AIK). Section 7.8.1 says "   When the first new Security Domain is created in a TEE for an SP, a new key pair is generated and associated with this SP. This key pair  is used ... instead of the device's TEE key pair. ". The AIK is used by the TAM and by the Client Application: 
 * The TAM uses the AIK to encrypt TA binary data and personalization data. 
 * The TA uses it to sign data sent to a Client Application (such as in the GetTAInformation call to query the TEE about installed TAs).

This again raises a couple of questions, namely 
 * Why does the TAM encrypt the TA binary when this should actually be done by the developer who creates the software and signs it? 
 * Is it necessary to provide anonymous attestation from the TA to the Client Application? 
 * Is it necessary to embed the optimized interaction between the Client Application and the TEE so deeply into the protocol? An alternative would be to route the interaction from the Client Application via the TAM to the Client Application, if a query about the installed TAs is actually needed. Normally, the TAM should have that information already anyway and wouldn't need to ask the TEE again.

**Comment by andyatyeo** [comment]
I am not necessarily against the idea of offering an implicit model where CreateSD is not used, but I do want to discuss what some consequences might be.
To address what value the SD currently provides:
**Additional Encryption of TA / perso data (by the TAM or the SP)**
 - The CreateSD response returned spaik public key. This could be used by either the TAM or the Service Provider (that sits behind the TAM) to encrypt a TA or personalization data, encrypted for that specific device. This provided a means where a ServiceProvider could be enabled to keep data (the TA and the perso data - which would typically contain sensitive per device info) hidden from the TAM. Is there value in this spaik? There are 2 cases:
 - 1. The TAM is fully trusted by all parties, and the Service Providers delegate all functionality to the TAM. If this is the case, then you might say that the JWE message level encryption of the InstallTA message is enough, without the need to have the additional spaik encryption. 
 - 2. (or) The TAM is of course trusted, but the Service Provider wants to maintain some independence and wants to keep the TA binary private from the TAM, and to be able to create secret ‘per device’ personalization data on the fly (that will be private from the TAM). An example might be that on a banking app, secret keys unique to a user need to be personalized for the TA, and it is not acceptable that any party except the device and the bank (SP) have any visibility of these keys.
 - Hannes mentioned that TA encryption could be implemented differently – in that the developer encrypts the TA (when they sign it). This would be an alternative approach, but it implies that that all TEEs must have the same (private or symmetric) decryption key that matches what the developer used to encrypt it. The question then is how this key gets injected into all the TEEs to enable the TEEs to decrypt the encrypted TA, and how this common key can be kept secret (since there would be so many copies of it).  I don’t think this approach would help with the TA personalization data

**getTAInformation response signed with the spaik**
 - Is there value in the anonymous attestation? Assuming the getTAInfo response is signed, there is value in avoiding releasing the TEE certificate to the app – since releasing the TEE certificate to rich OS apps implies that Rich OS can get a unique device identifier (which has privacy or tracking concerns). 
 - Hannes proposes an alternative flow where the TAM could be involved in the getTAInfo message (to avoid the need for the on device anonymous attestation). The downside of this is that it could stop the app from working offline (which would be undesirable)
 - If the CreateSD was removed, it wouldn’t necessarily stop us from being able to have an spaik for the purposes of signing getTAInfo response. (might need to think about this a bit more)
 - Perhaps it is worth thinking about what exactly this getTAInformation message is for. One case is for the RichOS to determine whether a TA is installed (so it can initiate an installation if it is not installed). However, there must be a TEE “Client” API that is out of scope of OTrP that enables the Rich OS to communicate with the TEE, and maybe this API might serve as a discovery mechanism as to whether a given TA is installed. Who should verify that the TA is trusted? It should be a function of the TEE itself – not the Rich OS to determine that the TA has not been tampered with. However – how does the Rich OS know that TA id xx that it is trying to talk to is the real TA and not an “imposter” that has been installed that happens to have the same TA id. Maybe we need to think more about how the Rich OS can be sure that TA id xx really does come from the service provider that it thinks it comes from.

**Comment by dmwheel1** [comment]
Andrew, thanks for the comments. A couple points that may indicate I need further clarification. 

If the primary purpose of the SD is to share an AIK for the device (your first bullet above), why not leverage the "getDeviceState" response? The dsi element (device state info) returned in the "getDeviceState" response already contains a list of the AIKs.

My confusion here is that if we are encrypting or integrity protecting things only to the device granularity, having multiple keys (one spaik per SD) seems over-engineered, or missing a requirement to warrant multiple keys. If only one key is required, then the "getDeviceState" response is sufficient.
Proof of source from a specific SP or specific TAM could be done with their key. Even when we have multiple spaik's, there is no cryptographic guarantee of proof-of-source from using those keys. What I mean is there is no benefit to the SP or TAM from having different spaiks to a device, since use of a particular spaik tied to a particular SD gives me no cryptographic binding, only some anecdotal separation. 

My question, or request for clarification, is why multiple spaiks? Why not just one per device?

If proof-of-source is necessary for the encrypted personalization data, the SP sending a symmetric key to the device, encrypted with the device's single spaik and signed by the SP seems more appropriate. This leads me to think I am missing something perhaps.

In general, when I think of an AIK, I think of the use between two parties, with the AIK being used to attest from the private key owner to the party that was (secretly) sent the public key. Not having the public key holder be able to send private data to the private key holder. Although, multiple such public keys affords some anonymity to the device, since a single public key cannot be traced to all transactions to the device. However, all the spaiks are found in the "DeviceStateInfo" which leaks significant info to the TAM. Thus, I am confused about the use of AIK here. 

For the getTAInfo interface for offline data from the TEE to a Client App, this interface would be provide by the TEEP Broker. The TEEP Broker could get a spaik from the TAM at some time, and this be able to retrieve info from the TEE. The problem occurs with where does the TEEP Broker keep that public key? The problem occurs when the TEEP Broker (or any component for that matter) on REE side wants to verify info from the TEE using the public key, how does the component on the REE side verify that the public key has not been replaced by an attacker? Verifying the data from the TEE without a certified public key becomes problematic, and affords no additional verification of the info from the TEE than if that data were not signed. 

I cannot get myself out of the circular logic here, but perhaps I am missing something or misunderstand some crucial aspect you are trying to communicate. I'd appreciate any clarification.

**Comment by andyatyeo** [comment]
To Davids questions:
**why not leverage getDeviceState response to return the spaik?**
It is true that getDeviceState can be used to return the spaik public key that would be required for encrypting TAbin/TAperso. The question is: what would trigger the generation of the spaik private key? Currently it is the CreateSD message that does this. If we issue an InstallTA without a CreateSD then at what point will that key get generated? (presumably if we went to a single spaik model this would be present on the TEE to begin with)

**why is there one spaik per service provider?**
(note technically it is one spaik per serviceprovider, and a SP could have multiple SDs in the TEE which would share the same spaik.) 
For the benefit of the SP?
My first reaction is that each SP wants their own key since it gives them a means to secretly communicate with that device that is secret from all other SPs (enc TA/TAbin). ... But as we know, the spaik is a RSA/ECC key and the SP has the public key. Therefore right now I cant think of a good reason that requires each SP to have their own unique SPAIK. (since a SP using the same public key to encrypt TA/TAbin would not allow another SP to decrypt it). 
For the benefit of the Rich OS app for verifying GetTAInfo?
If it is the same spaik, then multiple rich OS apps for differnt service providers could use the spaik publickey as a 'device fingerprint'. Whether this is a problem is another question.

I think you are right to question the logic here, and maybe we do need to rethink the question as to how the rich os app knows a specific TA is present

**Comment by dthaler** [comment]
Re "since releasing the TEE certificate to rich OS apps implies that Rich OS can get a unique device identifier (which has privacy or tracking concerns)".  Why is this an issue?  Rich OS's already have unique device identifiers, such as MAC addresses, IP addresses, and so forth.  The TEE certificate is just another one.  Since Rich OS's already have unique device identifiers, I don't see the problem in it getting another one.   Note this is independent from whether the Rich OS actually provides any of them to a Client Application (aka rich OS apps).  If the Rich OS doesn't provide any of them to a client app, then it could also provide an api that hides the TEE certificate and does verification itself at the app's request.

In other words, I don't yet understand the value in having an AIK separate from the TEE key.   I do believe there is value in each TA having its own key (and indeed the TCG's DICE mechanism can provide it, where the TA's cert chains up to the TEE's cert).

**Comment by mingpeiwk** [comment]
If we consider Rich OS cannot be fully trusted whereas we treat TEE and TA in higher protection, the unique ID protection by Rich OS may not be as strong. Currently TEEP Broken (in Rich OS) is considered to only relay encrypted messages between a TAM and a TEE.

Relying on Rich OS to hide the TEE certificate also adds a work into Rich OS. TEEP Broker isn't doing it today.

I think the main point circles back the privacy issue: is it OK for client apps to get TEE certificates? One  concern was that a client app may collect all TEE device certificates wherever it is installed.

**Comment by mingpeiwk** [comment]
On David W's comment about using one AIK per device instead of per SP, it is true that separate SP AIK doesn't offer cryptographic binding for the case by a SP to encrypt TA binary or TA personalization data. The same SPAIK was used for two use cases above.

If we can work around getTAInformation and privacy concern, we may reduce this, I think.

Implicit one SD per TA provides a simple model for some cases such as potentially IoT that doesn't need to explicitly create a SD.

Grouping TA by SD that a SP owns provide some small benefit for management when multiple TAs are supplied by a SP, such as deleting all TAs by one SP.

Some existing TEEs inherit a SD concept. Leaving SD completely out of TEEP will likely force future TEEs constraint how it uses a SD. We should be caution in breaking this "compatibility" as to completely removing SD concept.

**Comment by andyatyeo** [comment]
To respond to the comment previously "that Rich OS can already get many unique identifiers today, MAC address, IP address etc...". On desktop OS this is true, but on mobile OS unique identifiers are increasingly hidden (in order to prevent the ability of apps to track and correlate data). I understand that on the current major mobile releases, "device wide" identifiers are now hidden, and the unique identifiers available to apps are unique to that instance of an app rather than that device (to prevent aggregation of data). iOS has clamped down on this for some time and Android is moving in this direction. Therefore it feels as though we would be going against the flow to require that a per-device identifier would become available to the Rich OS

**Comment by dthaler** [comment]
I don’t follow.  The hiding mechanisms you’re talking about are Rich OS mechanisms that allow the Rich OS to use the unique identifiers without apps being able to use them.
So it’s not against that trend for the Rich OS to have them.

Indeed the Rich OS often uses them for things like another source of entropy in random number generating seeding, thus improving security.

**Comment by andyatyeo** [comment]
Dave - I agree that the hiding mechanisms mentioned do not stop the Rich OS itself from getting identifiers. Instead they stop individual apps (which should have restricted access) from getting identifiers for the device itself. (The identifiers increasingly being made available are unique to that instance of the app on that device, rather than the device itself). So a banking app may well want a unique identifier, but the identifier it sees is a different unique identifier that a shopping app from a different vendor on the same device sees. That approach (from the mobile OS vendors) is all about preventing apps from being able to aggregate data, and to 'sandbox' apps so they see their own data but aren't automatically allowed to see global properties of the device - e.g. IMEI number, MAC address etc.

**Comment by dthaler** [comment]
Although this issue was on the list of issues at IETF 105 that were supposedly ready to be closed, I checked the arch doc and it currently has this text:

5.8.  Trust Anchors in TEE

   Each TEE comes with a trust store that contains a whitelist of Trust
   Anchors that are used to validate a TAM's certificate.  A TEE will
   accept a TAM to create new Security Domains and install new TAs on
   behalf of an SP only if the TAM's certificate is chained to one of
   the root CA certificates in the TEE's trust store.
[…]

The term “Security Domains” is used above without any definition or reference to a definition.
If the GlobalPlatform-specific definition is intended as an example, then we could informatively point to GPD_SPE_120 (https://globalplatform.org/specs-library/tee-management-framework-including-asn1-profile/) section 1.4 for the definition (and elaboration in section 4.1).

5.13.  Security Domain
   No security domain (SD) is explicitly assumed in a TEE for TA
   management.  Some TEE, for example, some TEE compliant with Global
   Platform (GP), may continue to choose to use SD to organize resource
   partition and security boundaries.  It is up to a TEE implementation
   to decide how a SD is attached to a TA installation, for example, one
   SD could be created per TA.

Was “explicitly” meant to be “implicitly”?   (Also there's some minor grammatical issues with the text.)

So unless someone can convince me why the above text is all correct, I’m reluctant to close this issue.

**Comment by dthaler** [comment]
@ncamwing I believe this is now ready to be closed, as PR #72 is now merged
